### PR TITLE
Improved Automation tests on Operator, pipeline spec

### PR DESCRIPTION
**Pull request by praveen4g0**
- Operator installation by subscribing to canary channel on real cluster.
- Operator deletion from cluster
- Creating simple pipeline and run pipeline using tkn client

**Assumptions:**

- latest oc binary should be installed.
- latest tkn binary which is compatible with operator should be installed.

### CODE REVIEW

File: pkg/client/clients.go [hunk 1]
```diff
@@ -0,0 +1,199 @@
+package client
```
Comment (sthaha) [inline pkg/client/clients.go]: @praveen4g0 , if the file is copied from `pipelines`, can't we not use it through `go modules`/`vendor` instead of making a copy?

File: pkg/helper/helper.go [hunk 2]
```diff
@@ -0,0 +1,251 @@
+package helper
```
Comment (sthaha) [inline pkg/helper/helper.go]: Did you try adding operator as a  dependency?

File: pkg/operator/install.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+	"github.com/openshift-pipelines/release-tests/pkg/helper"
+	. "github.com/smartystreets/goconvey/convey"
+	"github.com/tektoncd/operator/pkg/apis/operator/v1alpha1"
+	op "github.com/tektoncd/operator/pkg/apis/operator/v1alpha1"
+	"github.com/tektoncd/pipeline/pkg/names"
+	"gotest.tools/v3/icmd"
+	knativetest "knative.dev/pkg/test"
+)
+
+var CR *v1alpha1.Config
```
Comment (sthaha) [inline pkg/operator/install.go]: how about descriptive names for exported variables ; e.g.
`Config`, `Clients`

File: pkg/operator/install.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+var Clt *client.Clients
+var t = &testing.T{}
```
Comment (sthaha) [inline pkg/operator/install.go]: how is this used?

File: pkg/pipelines/pipelines.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+	"time"
+
+	"github.com/openshift-pipelines/release-tests/pkg/client"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
+	tb "github.com/tektoncd/pipeline/test/builder"
+	"gotest.tools/v3/icmd"
+)
+
+const (
+	TaskName1                 = "create-file"
```
Comment (sthaha) [inline pkg/pipelines/pipelines.go]: Can we have a better name  ( e.g. 'CreateFileTask` ) than `TaskName1` especially because it is exported?

File: pkg/pipelines/pipelines.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+	TaskName2                 = "check-stuff-file-exists"
+	tePipelineName            = "output-pipeline"
+	tePipelineGitResourceName = "skaffold-git"
+)
+
+func getGitResourceForOutPutPipeline(rname string, namespace string) *v1alpha1.PipelineResource {
```
Comment (sthaha) [inline pkg/pipelines/pipelines.go]: nit: `OutPut` -> `Output`

Comment (sthaha) [inline pkg/pipelines/pipelines.go]: how about `newGitResource` ?

File: pkg/pipelines/pipelines.go [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+	return tb.PipelineResource(rname, namespace, tb.PipelineResourceSpec(
+		v1alpha1.PipelineResourceTypeGit,
+		tb.PipelineResourceSpecParam("url", "https://github.com/GoogleContainerTools/skaffold"),
+		tb.PipelineResourceSpecParam("revision", "master"),
+	))
+}
+
+func getCreateFileTask(taskname string, namespace string) *v1alpha1.Task {
```
Comment (sthaha) [inline pkg/pipelines/pipelines.go]: how about `newCreateFileTask` ?

File: pkg/pipelines/pipelines.go [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+		tb.TaskInputs(tb.InputsResource("workspace", v1alpha1.PipelineResourceTypeGit, tb.ResourceTargetPath("damnworkspace"))),
+		tb.TaskOutputs(tb.OutputsResource("workspace", v1alpha1.PipelineResourceTypeGit)),
+		tb.Step("read-docs-old", "ubuntu", tb.StepCommand("/bin/bash"), tb.StepArgs("-c", "ls -la /workspace/damnworkspace/docs/README.md")),
+		tb.Step("write-new-stuff", "ubuntu", tb.StepCommand("bash"), tb.StepArgs("-c", "ln -s /workspace/damnworkspace /workspace/output/workspace && echo some stuff > /workspace/output/workspace/stuff")),
+	}
+
+	return tb.Task(taskname, namespace, tb.TaskSpec(taskSpecOps...))
+}
+
+func getReadFileTask(taskname string, namespace string) *v1alpha1.Task {
```
Comment (sthaha) [inline pkg/pipelines/pipelines.go]: newReadFileTask ?

File: pkg/pipelines/pipelines.go [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+
+	taskSpecOps := []tb.TaskSpecOp{
+		tb.TaskInputs(tb.InputsResource("workspace", v1alpha1.PipelineResourceTypeGit, tb.ResourceTargetPath("newworkspace"))),
+		tb.Step("read", "ubuntu", tb.StepCommand("/bin/bash"), tb.StepArgs("-c", "cat /workspace/newworkspace/stuff")),
+	}
+
+	return tb.Task(taskname, namespace, tb.TaskSpec(taskSpecOps...))
+}
+
+func getPipeline(pipelineName string, namespace string, createFiletaskName string, readFileTaskName string) *v1alpha1.Pipeline {
```
Comment (sthaha) [inline pkg/pipelines/pipelines.go]: how about `newCreateAndReadFilePipeline` ?

Comment (sthaha) [reply pkg/pipelines/pipelines.go]: Does it really need to accept the names of the tasks?

File: pkg/pipelines/pipelines.go [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+	if _, err := c.TaskClient.Create(getCreateFileTask(TaskName1, namespace)); err != nil {
+		log.Fatalf("Failed to create Task Resource `%s`: %s", TaskName1, err)
+	}
+
+	log.Printf("Creating Task  %s", TaskName2)
+	if _, err := c.TaskClient.Create(getReadFileTask(TaskName2, namespace)); err != nil {
+		log.Fatalf("Failed to create Task Resource `%s`: %s", TaskName2, err)
+	}
+
+	log.Printf("Create Pipeline %s", tePipelineName)
```
Comment (sthaha) [inline pkg/pipelines/pipelines.go]: what does `te` in `tePipelineName` mean?

File: pkg/pipelines/pipelines.go [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+	if _, err := c.PipelineClient.Create(getPipeline(tePipelineName, namespace, TaskName1, TaskName2)); err != nil {
+		log.Fatalf("Failed to create pipeline `%s`: %s", tePipelineName, err)
+	}
+
+	time.Sleep(1 * time.Second)
```
Comment (sthaha) [inline pkg/pipelines/pipelines.go]: why is this needed?

File: pkg/client/clients.go (hunk 1 above)
Comment (praveen4g0) [reply pkg/client/clients.go]: @sthaha No it's not Copy of Clients from ```Pipelines``` , and Also ```clients``` from Pipelines are not exportable, Reason I have come up with this clients.go file is, to install an operator, validate it on any Cluster (OCP), instead of using ```operator-sdk local```

File: pkg/helper/helper.go (hunk 2 above)
Comment (praveen4g0) [reply pkg/helper/helper.go]: Yes, I have used ```github.com/tektoncd/operator/pkg/apis/operator/v1alpha1``` config schema to validated cluster Resource Creation & Deletion

File: pkg/operator/install.go (hunk 3 above)
Comment (praveen4g0) [reply pkg/operator/install.go]: That makes sense Really, I am updating it to ```ClusterConfig``` , ```ClientSet```

File: pkg/helper/helper.go [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+	"log"
+	"testing"
+	"time"
+
+	"github.com/openshift-pipelines/release-tests/pkg/client"
+	"github.com/openshift-pipelines/release-tests/pkg/config"
+	secv1 "github.com/openshift/api/security/v1"
+	secclient "github.com/openshift/client-go/security/clientset/versioned/typed/security/v1"
+	"github.com/tektoncd/operator/pkg/apis/operator/v1alpha1"
+	op "github.com/tektoncd/operator/pkg/apis/operator/v1alpha1"
```
Comment (ppitonak) [inline pkg/helper/helper.go]: Why twice?

File: go.mod [hunk 13]
```diff
@@ -1,5 +1,37 @@
...
+	k8s.io/api v0.0.0-20190612125737-db0771252981
+	k8s.io/apiextensions-apiserver v0.0.0-20190820104113-47893d27d7f7
+	k8s.io/apimachinery v0.0.0-20190612125636-6a5db36e93ad
+	k8s.io/client-go v11.0.0+incompatible
+	knative.dev/pkg v0.0.0-20191216211902-b26ddf762bc9
+	sigs.k8s.io/controller-runtime v0.1.12
+)
 
-require github.com/smartystreets/goconvey v1.6.4
+replace (
```
Comment (ppitonak) [inline go.mod]: 1. why two `replace` sections?
2. why to use `replace` at all?

Comment (praveen4g0) [reply go.mod]: I have updated go.mod file in #7

**Comment by praveen4g0** [comment]
Closing this PR #6 As I have raised a new PR  #7, with Ginkgo framework

File: pkg/helper/helper.go (hunk 12 above)
Comment (praveen4g0) [reply pkg/helper/helper.go]: ahh.. Good catch will update that.
